File: 503.下一个更大元素-ii.c
/*
 * @lc app=leetcode.cn id=503 lang=c
 *
 * [503] 下一个更大元素 II
 */

// @lc code=start

/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
int *nextGreaterElements(int *nums, int numsSize, int *returnSize)
{
  int i, top, *res, *stack;
  stack = (int *)malloc(sizeof(int) * 2 * numsSize);
  res = (int *)malloc(sizeof(int) * numsSize);
  for (i = 0; i < numsSize; i++)
    res[i] = -1;
  top = 0;
  stack[top++] = 0;

  for (i = 1; i < 2 * numsSize; i++)
  {
    while (top != 0 && nums[stack[top - 1]] < nums[i % numsSize])
    {
      res[stack[top - 1]] = nums[i % numsSize];
      top--;
    }
    stack[top++] = i % numsSize;
  }
  *returnSize = numsSize;
  return res;
}
// @lc code=end

File: store/503.下一个更大元素-ii.js
/*
 * @lc app=leetcode.cn id=503 lang=javascript
 *
 * [503] 下一个更大元素 II
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var nextGreaterElements = function (nums) {
  let res = new Array(nums.length * 2).fill(-1),
    stack = [],
    top = 0;

  nums = [...nums, ...nums];

  stack[top++] = 0;
  for (let i = 1, len = nums.length; i < len; i++) {
    while (nums[i] > nums[stack[top - 1]] && top !== 0) {
      res[stack[top - 1]] = nums[i];
      top--;
    }
    stack[top++] = i;
  }

  return res.slice(0, nums.length / 2);
};
// @lc code=end
